File: binary_patcher.py
# will this work? maybe, we can find out. i know the c# version of this did work before with this.
import os
import shutil

CORE_PAIRS = [
    ("messenger.hotmail.com", "ms.msgrsvcs.ctsrv.gay"),
    ("nexus.passport.com", "pp.login.ugnet.gay")
]

GATEWAY_PAIR = ("gateway.messenger.hotmail.com", "httpgws.ms.msgrsvcs.ctsrv.gay")

def patch_binary(target_file, include_gateway):
    """Hex patches the executable's ASCII server strings in-place."""
    if not os.path.exists(target_file):
        raise FileNotFoundError(f"Target file not found: {target_file}")

    # Create backup
    backup = target_file + ".bak"
    if not os.path.exists(backup):
        shutil.copyfile(target_file, backup)
        print(f"Backup saved to {backup}")
    else:
        print(f"Backup already exists at {backup} (not overwriting)")

    pairs = list(CORE_PAIRS)
    if include_gateway:
        pairs.append(GATEWAY_PAIR)

    with open(target_file, "rb") as f:
        data = bytearray(f.read())

    total_replacements = 0

    for find_str, replace_str in pairs:
        find_bytes = find_str.encode("ascii")
        replace_bytes = replace_str.encode("ascii")

        if len(replace_bytes) > len(find_bytes):
            print(f"Skipping '{find_str}' -> '{replace_str}': replacement is longer than original.")
            continue

        # If replacement is shorter, pad with null bytes (0x00)
        if len(replace_bytes) < len(find_bytes):
            replace_bytes = replace_bytes.ljust(len(find_bytes), b'\x00')

        occurrences = 0
        index = 0
        while True:
            index = data.find(find_bytes, index)
            if index == -1:
                break
            data[index:index+len(find_bytes)] = replace_bytes
            occurrences += 1
            index += len(find_bytes)

        print(f"'{find_str}' -> '{replace_str}': {occurrences} occurrence(s) patched")
        total_replacements += occurrences

    if total_replacements > 0:
        with open(target_file, "wb") as f:
            f.write(data)
        print(f"Done. {total_replacements} total replacement(s) written to {target_file}")
    else:
        print("No matching strings were found. Wrong file/version?")

File: BinaryPatcher.cs
using System.Text;

namespace CrossTalkPatcher;

/// <summary>
///  makes the ascii strings point to the actual servers, if you want to know why, go to https://crosstalk.im/guides/patching and look at MSN/Windows Messenger 4.7.2009 - 3001 patching guide with Hex
/// </summary>
public static class BinaryPatcher
{
    public record StringPair(string Find, string Replace);

    public static readonly StringPair[] CorePairs =
    {
        new("messenger.hotmail.com", "ms.msgrsvcs.ctsrv.gay"),
        new("nexus.passport.com", "pp.login.ugnet.gay"),
    };

    public static readonly StringPair GatewayPair =
        new("gateway.messenger.hotmail.com", "httpgws.ms.msgrsvcs.ctsrv.gay");

    public static void Patch(string targetFile, bool includeGateway)
    {
        if (!File.Exists(targetFile))
            throw new FileNotFoundException("Target file not found", targetFile);

        string backup = targetFile + ".bak";
        if (!File.Exists(backup))
        {
            File.Copy(targetFile, backup);
            Console.WriteLine($"Backup saved to {backup}");
        }
        else
        {
            Console.WriteLine($"Backup already exists at {backup} (not overwriting)");
        }

        var pairs = new List<StringPair>(CorePairs);
        if (includeGateway) pairs.Add(GatewayPair);

        byte[] bytes = File.ReadAllBytes(targetFile);
        int totalReplacements = 0;

        foreach (var pair in pairs)
        {
            byte[] findBytes = Encoding.ASCII.GetBytes(pair.Find);
            byte[] replaceBytes = Encoding.ASCII.GetBytes(pair.Replace);

            if (replaceBytes.Length > findBytes.Length)
            {
                Console.WriteLine($"Skipping '{pair.Find}' -> '{pair.Replace}': replacement is longer than original.");
                continue;
            }

            if (replaceBytes.Length < findBytes.Length)
            {
                var padded = new byte[findBytes.Length]; // defaults to 0x00
                Array.Copy(replaceBytes, padded, replaceBytes.Length);
                replaceBytes = padded;
            }

            int occurrences = 0;
            for (int i = 0; i <= bytes.Length - findBytes.Length; i++)
            {
                bool match = true;
                for (int j = 0; j < findBytes.Length; j++)
                {
                    if (bytes[i + j] != findBytes[j]) { match = false; break; }
                }
                if (match)
                {
                    Array.Copy(replaceBytes, 0, bytes, i, replaceBytes.Length);
                    occurrences++;
                    i += findBytes.Length - 1;
                }
            }

            Console.WriteLine($"'{pair.Find}' -> '{pair.Replace}': {occurrences} occurrence(s) patched");
            totalReplacements += occurrences;
        }

        if (totalReplacements > 0)
        {
            File.WriteAllBytes(targetFile, bytes);
            Console.WriteLine($"Done. {totalReplacements} total replacement(s) written to {targetFile}");
        }
        else
        {
            Console.WriteLine("No matching strings were found. Wrong file/version?");
        }
    }
}